### PR TITLE
SONARJAVA-2390 S2187 Support JUnit 5 Testable meta-annotation

### PR DESCRIPTION
**Pull request by saberduck**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/NoTestInTestClassCheck.java [hunk 1]
```diff
@@ -38,21 +39,46 @@
...
 import java.util.stream.Stream;
 
 @Rule(key = "S2187")
 public class NoTestInTestClassCheck extends IssuableSubscriptionVisitor {
 
-  private static final Predicate<SymbolMetadata.AnnotationInstance> PREDICATE_ANNOTATION_TEST_OR_UNKNOWN = input -> {
-    Type type = input.symbol().type();
-    return type.isUnknown() || isTestAnnotation(type);
-  };
+  private final Map<String, Boolean> testAnnotations;
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/NoTestInTestClassCheck.java]: As discussed : while unlikely to happen in this case , this is a memory leak in disguise and goes against the usual way to write checks that are running on a file by file basis.

File: java-checks/src/main/java/org/sonar/java/checks/NoTestInTestClassCheck.java [hunk 2]
```diff
@@ -38,21 +39,46 @@
...
+
+  public NoTestInTestClassCheck() {
+    testAnnotations = new HashMap<>();
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/NoTestInTestClassCheck.java]: This map should probably be replaced by two sets : "encouteredAnnotations" and "testableAnnotations" : this should make the code simpler in the end.

File: java-checks/src/main/java/org/sonar/java/checks/NoTestInTestClassCheck.java [hunk 3]
```diff
@@ -38,21 +38,38 @@
...
-      || type.symbol().metadata().isAnnotatedWith("org.junit.jupiter.api.Test");
+  private boolean isTestAnnotation(Type type) {
+    return testAnnotations.contains(type.fullyQualifiedName()) || isJUnitTestableMetaAnnotated(type);
+  }
+
+  private boolean isJUnitTestableMetaAnnotated(Type type) {
+    if (seenAnnotations.contains(type.fullyQualifiedName())) {
+      return false;
+    }
+    seenAnnotations.add(type.fullyQualifiedName());
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/NoTestInTestClassCheck.java]: You can skip contains call by relying on result of add.

Comment (saberduck) [inline java-checks/src/main/java/org/sonar/java/checks/NoTestInTestClassCheck.java]: I am not really big fan of using `add` that way, I find it easier to read, when there are no side-effects in conditions.

**Pull request merged**